src/main.rs

```diff
@@ -1,3 +1,146 @@
+        ),
+        // TODO better error handling
+        _ => panic!("Maker initialisation failed!"),
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    // TODO: from config or hardcoded
+    let dai_contract_addr: comit::ethereum::Address = unimplemented!();
+
+    // TODO: Proper wallet initialisation from config
+    let bitcoin_wallet =
+        bitcoin_wallet::Wallet::new(unimplemented!(), unimplemented!(), unimplemented!()).unwrap();
+    let ethereum_wallet =
+        ethereum_wallet::Wallet::new(unimplemented!(), unimplemented!(), unimplemented!()).unwrap();
+
+    let maker = init_maker(bitcoin_wallet, ethereum_wallet).await;
+
+    let orderbook = Orderbook;
+    let nectar = Nectar::new(orderbook);
+
+    let mut swarm: libp2p::Swarm<Nectar> = unimplemented!();
+
+    let initial_sell_order = maker.next_sell_order();
+    let initial_buy_order = maker.next_buy_order();
+
+    match (initial_sell_order, initial_buy_order) {
+        (Ok(sell_order), Ok(buy_order)) => {
+            swarm.orderbook.publish(sell_order.into());
+            swarm.orderbook.publish(buy_order.into());
+        }
+        _ => panic!("Unable to publish initial orders!"),
+    }
+
+    loop {
+        let rate = get_btc_dai_mid_market_rate().await;
+        match rate {
+            Ok(new_rate) => maker.update_rate(unimplemented!()), // maker should record timestamp of this
+            Err(e) => maker.track_failed_rate(e),
+        }
+
+        let bitcoin_balance = bitcoin_wallet.balance().await;
+        match bitcoin_balance {
+            Ok(new_balance) => maker.update_bitcoin_balance(new_balance), // maker should record timestamp of this
+            Err(e) => maker.track_failed_balance_update(e),
+        }
+        let dai_balance = ethereum_wallet.dai_balance().await;
+        match dai_balance {
+            Ok(new_balance) => maker.update_dai_balance(new_balance.into()), // maker should record timestamp of this
+            Err(e) => maker.track_failed_balance_update(e),
+        }
+
+        // if nothing happens on the network for 15 seconds, loop
+        // again
+
+        // ASSUMPTION: a BtcDaiOrder Sell order and a BtcDaiOrder Buy
+        // order were published before we enter the loop (during
+        // initialization)
+        #[allow(clippy::single_match)]
+        match tokio::time::timeout(Duration::from_secs(15), swarm.next()).await {
+            Ok(event) => {
+                match event {
+                    network::Event::OrderTakeRequest(order) => {
+                        // decide & take & reserve
+                        let reaction = maker.react_to_taken_order(order.clone());
+
+                        match reaction {
+                            Ok(Reaction::Confirmed { next_order }) => {
+                                swarm.orderbook.take(order);
+                                orderbook.publish(next_order.into());
+                            }
+                            Ok(Reaction::RateSucks)
```

Comment:
If we handle all these scenarios the same way then it is a hint that you do not need to differentiate them. If the maker does not execute the order then it is ok to return an error instead of disguising it in an `Ok` result.